datastructs/ring.c

```diff
@@ -0,0 +1,76 @@
+// ring.c : 
+/* A ring buffer for multi-threaded programs. A single producer pushes
+* objects into the ring, while a single consumer pops them off for
+* processing. If the ring is full (i.e., the consumer is slower than the
+* producer), then the push operation fails.
+*/
+
+
+
+
+#include <stdlib.h>
+#include <stdio.h>
+
+#include "ring.h"
+
+
+
+
+/* Create a ring buffer with the specified size. Return the ring or NULL
+if there is not enough memory to create. */
+struct ring* ring_create(int size)
+{
+	if (size <= 0)  return NULL;
+
+
+
+	struct ring *rb = (struct ring*) malloc( sizeof(struct ring) );
+	rb->push = 0;
+	rb->pop = 0;
+
+	rb->size = size;
+	rb->count = 0;
+
+	rb->buffer = (void*) malloc( size*sizeof(int) );
```

Comment:
while this code will generally work, you don't really want buffer to be sized based on `ints`, you want it to be based on the size of a `void*` (i.e., a generic pointer).